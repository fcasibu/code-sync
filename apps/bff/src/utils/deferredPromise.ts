export class DeferredPromise<T> {
  public promise: Promise<T>;

  public resolve: (value: T | PromiseLike<T>) => void = () => undefined;
  public reject: (reason?: unknown) => void = () => undefined;

  constructor() {
    this.promise = new Promise((res, rej) => {
      this.resolve = res;
      this.reject = rej;
    });
  }
}
