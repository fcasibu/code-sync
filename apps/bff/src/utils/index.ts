export * from './urlForHTTPServer';
export * from './gqlHelper';
export * from './testHelpers';
export * from './lazy';
export * from './typeGuards';
