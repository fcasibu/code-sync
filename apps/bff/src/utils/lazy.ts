export const lazy = <T>(factory: () => T) => {
  let instance: T;

  return () => {
    if (!instance) {
      instance = factory();
    }

    return instance;
  };
};
