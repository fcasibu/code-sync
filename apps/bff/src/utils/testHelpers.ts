import type { PrismaClient } from '@code-sync/db';
import { prismaClient } from '../config';
import type { Application } from '../server';
import { startServer } from '../server';

type UnMaybe<T> = {
  [P in keyof T]-?: NonNullable<T[P]>;
};

export const testHelper = () => {
  let _app: Application | undefined;
  let _prismaClient: PrismaClient | undefined;

  const setup = async () => {
    _prismaClient = prismaClient();
    _app = await startServer({ port: 0, prismaClient: _prismaClient });
  };

  const teardown = async () => {
    await _app?.apolloServer?.stop();
    await _prismaClient?.$disconnect();
    _app?.expressServer?.close();
  };

  return {
    setup,
    teardown,
    get app() {
      if (!_app)
        throw new Error(
          '`setup` in the testHelper was not invoked. Invoke the `setup` of `testHelper` in either `beforeEach` or `beforeAll`',
        );

      return _app as UnMaybe<Application>;
    },
    get prismaClient() {
      if (!_prismaClient)
        throw new Error(
          '`setup` in the testHelper was not invoked. Invoke the `setup` of `testHelper` in either `beforeEach` or `beforeAll`',
        );

      return _prismaClient;
    },
  };
};
