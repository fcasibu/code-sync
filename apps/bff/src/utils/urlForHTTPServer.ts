import type { Server } from 'node:http';
import type { AddressInfo } from 'node:net';
import { format } from 'node:url';

// https://github.com/apollographql/apollo-server/blob/main/packages/server/src/utils/urlForHttpServer.ts
export function urlForHTTPServer(expressServer: Server) {
  const { address, port } = expressServer.address() as AddressInfo;

  // Convert IPs which mean "any address" (IPv4 or IPv6) into localhost
  // corresponding loopback ip. Note that the url field we're setting is
  // primarily for consumption by our test suite. If this heuristic is wrong for
  // your use case, explicitly specify a frontend host (in the `host` option
  // when listening).
  const hostname = address === '' || address === '::' ? 'localhost' : address;

  return format({
    protocol: 'http',
    hostname,
    port,
    pathname: '/',
  });
}
