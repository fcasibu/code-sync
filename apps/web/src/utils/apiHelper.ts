import { type NextRequest, NextResponse } from 'next/server';
import { INTERNAL_HEADER_API_KEY } from '@/constants';
import { env } from '@/env';

export const isAuthorized = (req: NextRequest) =>
  req.headers.get(INTERNAL_HEADER_API_KEY) ===
  env.INTERNAL_HEADER_API_KEY_VALUE;

export const sendUnauthorizedResponse = () =>
  NextResponse.json({ error: 'Unauthorized' }, { status: 401 });
