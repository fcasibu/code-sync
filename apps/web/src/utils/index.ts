export * from './gqlHelper';
export * from './runtime';
export * from './apiHelper';
