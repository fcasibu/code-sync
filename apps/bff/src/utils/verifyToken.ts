import { promisify } from 'node:util';
import { strict as assert } from 'assert';
import type { JwtPayload } from 'jsonwebtoken';
import jwt from 'jsonwebtoken';
import { JwksClient } from 'jwks-rsa';
import { logger } from '@code-sync/logger';
import { DeferredPromise } from './deferredPromise';

export const verifyToken = (token: string) => {
  const client = new JwksClient({
    jwksUri: `https://${process.env.AUTH0_DOMAIN}/.well-known/jwks.json`,
  });

  const deferredPromise = new DeferredPromise<jwt.JwtPayload>();

  const getSigningKey = promisify(client.getSigningKey);

  const getJwksClientKey = async (
    header: jwt.JwtHeader,
    callback: jwt.SigningKeyCallback,
  ) => {
    const key = await getSigningKey(header.kid);
    callback(null, key?.getPublicKey());
  };

  jwt.verify(
    token,
    getJwksClientKey,
    getJwtVerifyOptions(),
    (error, decoded) => {
      if (error) {
        logger.error(error, 'Was unable to verify token');
        deferredPromise.resolve({});
        return;
      }

      deferredPromise.resolve(decoded as JwtPayload);
    },
  );

  return deferredPromise.promise;
};

const getJwtVerifyOptions = (): jwt.VerifyOptions => {
  const { AUTH0_AUDIENCE, AUTH0_DOMAIN } = process.env;

  assert(AUTH0_AUDIENCE, 'AUTH0_AUDIENCE is required environment variable.');
  assert(AUTH0_DOMAIN, 'AUTH0_DOMAIN is required environment variable.');

  return {
    audience: AUTH0_AUDIENCE,
    issuer: `https://${AUTH0_DOMAIN}/`,
    algorithms: ['RS256'],
  };
};
