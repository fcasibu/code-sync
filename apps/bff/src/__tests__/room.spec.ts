import type { CreateRoomInput } from 'src/graphql/types';
import requestGQL from 'supertest-graphql';
import { afterAll, beforeAll, describe, expect, it } from 'vitest';
import type { ResultOf, VariablesOf } from '../utils/gqlHelper';
import { graphql } from '../utils/gqlHelper';
import { testHelper } from '../utils/testHelpers';

const helper = testHelper();

beforeAll(helper.setup);
afterAll(helper.teardown);

describe('resolver#Room', () => {
  describe('rooms', () => {
    it('should return a list of rooms', async () => {
      const query = graphql(`
        query RoomQuery {
          rooms {
            id
            name
            description
            document {
              id
              content
              language
            }
            owner {
              id
              displayName
              profilePicture
            }
            spectators {
              id
              name
              canEdit
              isGuest
            }
          }
        }
      `);

      const response = await requestGQL<ResultOf<typeof query>>(
        helper.app.expressServer,
      )
        .query(query)
        .expectNoErrors();

      expect(response.data?.rooms).not.toHaveLength(0);
      expect(response.data).toMatchSnapshot();
    });
  });

  describe('room', () => {
    it('should return a room through their ID', async () => {
      const query = graphql(`
        query RoomQuery($input: RoomByIdInput!) {
          room(input: $input) {
            id
            name
            description
            document {
              id
              content
              language
            }
            owner {
              id
              displayName
              profilePicture
            }
            spectators {
              id
              name
              canEdit
              isGuest
            }
          }
        }
      `);

      const roomId = 'room-0';
      const response = await requestGQL<
        ResultOf<typeof query>,
        VariablesOf<typeof query>
      >(helper.app.expressServer)
        .query(query)
        .variables({ input: { id: roomId } })
        .expectNoErrors();

      expect(response.data?.room?.id).toBe(roomId);
      expect(response.data).toMatchSnapshot();
    });
  });

  describe('createRoom', () => {
    it('should create a room and return it', async () => {
      const mutation = graphql(`
        mutation RoomMutation($input: CreateRoomInput!) {
          createRoom(input: $input) {
            id
            name
            description
            document {
              id
              content
              language
            }
            owner {
              id
              displayName
              profilePicture
            }
            spectators {
              id
              name
              canEdit
              isGuest
            }
          }
        }
      `);

      const document = await helper.prismaClient.document.create({
        data: {
          language: 'TYPESCRIPT',
          content: "console.log('Hello, World!');",
        },
      });

      const ownerId = 'user-0';
      const input: CreateRoomInput = {
        ownerId,
        documentId: document.id,
        name: 'test room',
        description: 'This is a test room',
      };
      const response = await requestGQL<
        ResultOf<typeof mutation>,
        VariablesOf<typeof mutation>
      >(helper.app.expressServer)
        .mutate(mutation)
        .variables({
          input,
        })
        .expectNoErrors();

      expect(response.data?.createRoom?.document.language).toMatch(
        document.language,
      );
      expect(response.data?.createRoom?.document.content).toMatch(
        document.content,
      );
      expect(response.data?.createRoom?.document.id).toBe(document.id);
      expect(response.data?.createRoom?.owner.id).toMatch(/0/i);
      expect(response.data?.createRoom?.name).toMatch(/test room/i);
    });
  });
});
