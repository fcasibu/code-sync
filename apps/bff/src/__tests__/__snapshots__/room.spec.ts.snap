// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`resolver#Room > room > should return a room through their ID 1`] = `
{
  "room": {
    "description": null,
    "document": {
      "content": "",
      "id": "document-0",
      "language": "TYPESCRIPT",
    },
    "id": "room-0",
    "name": "Optimized maximized paradigm",
    "owner": {
      "displayName": "Katharina",
      "id": "user-0",
      "profilePicture": "https://via.placeholder.com/150x150/bc9de8/96d1f5.jpeg?text=Katharina",
    },
    "spectators": [
      {
        "canEdit": false,
        "id": "spectator-0",
        "isGuest": true,
        "name": "Katharina_Rau85",
      },
      {
        "canEdit": false,
        "id": "spectator-1",
        "isGuest": true,
        "name": "Hayley12",
      },
      {
        "canEdit": false,
        "id": "spectator-2",
        "isGuest": true,
        "name": "Hudson_Balistreri94",
      },
      {
        "canEdit": false,
        "id": "spectator-3",
        "isGuest": true,
        "name": "Kathryne.Quigley12",
      },
      {
        "canEdit": false,
        "id": "spectator-4",
        "isGuest": true,
        "name": "Vito97",
      },
      {
        "canEdit": false,
        "id": "spectator-5",
        "isGuest": true,
        "name": "Damaris.Strosin36",
      },
      {
        "canEdit": false,
        "id": "spectator-6",
        "isGuest": true,
        "name": "Sister82",
      },
      {
        "canEdit": false,
        "id": "spectator-7",
        "isGuest": true,
        "name": "Arden43",
      },
      {
        "canEdit": false,
        "id": "spectator-8",
        "isGuest": true,
        "name": "Schuyler86",
      },
      {
        "canEdit": false,
        "id": "spectator-9",
        "isGuest": true,
        "name": "Adell.Kuvalis",
      },
      {
        "canEdit": false,
        "id": "spectator-10",
        "isGuest": true,
        "name": "Olen_Aufderhar",
      },
      {
        "canEdit": false,
        "id": "spectator-11",
        "isGuest": true,
        "name": "Chelsie.Aufderhar",
      },
      {
        "canEdit": false,
        "id": "spectator-12",
        "isGuest": true,
        "name": "Cara.Robel8",
      },
      {
        "canEdit": false,
        "id": "spectator-13",
        "isGuest": true,
        "name": "Orion_Ferry89",
      },
      {
        "canEdit": false,
        "id": "spectator-14",
        "isGuest": true,
        "name": "Jordan_Rutherford",
      },
      {
        "canEdit": false,
        "id": "spectator-15",
        "isGuest": true,
        "name": "Rosie.Daugherty",
      },
      {
        "canEdit": false,
        "id": "spectator-16",
        "isGuest": true,
        "name": "Damien55",
      },
      {
        "canEdit": false,
        "id": "spectator-17",
        "isGuest": true,
        "name": "Eleanora.Lehner88",
      },
      {
        "canEdit": false,
        "id": "spectator-18",
        "isGuest": true,
        "name": "Mable_Lakin",
      },
      {
        "canEdit": false,
        "id": "spectator-19",
        "isGuest": true,
        "name": "Axel.Ruecker",
      },
    ],
  },
}
`;

exports[`resolver#Room > rooms > should return a list of rooms 1`] = `
{
  "rooms": [
    {
      "description": null,
      "document": {
        "content": "",
        "id": "document-0",
        "language": "TYPESCRIPT",
      },
      "id": "room-0",
      "name": "Optimized maximized paradigm",
      "owner": {
        "displayName": "Katharina",
        "id": "user-0",
        "profilePicture": "https://via.placeholder.com/150x150/bc9de8/96d1f5.jpeg?text=Katharina",
      },
      "spectators": [
        {
          "canEdit": false,
          "id": "spectator-0",
          "isGuest": true,
          "name": "Katharina_Rau85",
        },
        {
          "canEdit": false,
          "id": "spectator-1",
          "isGuest": true,
          "name": "Hayley12",
        },
        {
          "canEdit": false,
          "id": "spectator-2",
          "isGuest": true,
          "name": "Hudson_Balistreri94",
        },
        {
          "canEdit": false,
          "id": "spectator-3",
          "isGuest": true,
          "name": "Kathryne.Quigley12",
        },
        {
          "canEdit": false,
          "id": "spectator-4",
          "isGuest": true,
          "name": "Vito97",
        },
        {
          "canEdit": false,
          "id": "spectator-5",
          "isGuest": true,
          "name": "Damaris.Strosin36",
        },
        {
          "canEdit": false,
          "id": "spectator-6",
          "isGuest": true,
          "name": "Sister82",
        },
        {
          "canEdit": false,
          "id": "spectator-7",
          "isGuest": true,
          "name": "Arden43",
        },
        {
          "canEdit": false,
          "id": "spectator-8",
          "isGuest": true,
          "name": "Schuyler86",
        },
        {
          "canEdit": false,
          "id": "spectator-9",
          "isGuest": true,
          "name": "Adell.Kuvalis",
        },
        {
          "canEdit": false,
          "id": "spectator-10",
          "isGuest": true,
          "name": "Olen_Aufderhar",
        },
        {
          "canEdit": false,
          "id": "spectator-11",
          "isGuest": true,
          "name": "Chelsie.Aufderhar",
        },
        {
          "canEdit": false,
          "id": "spectator-12",
          "isGuest": true,
          "name": "Cara.Robel8",
        },
        {
          "canEdit": false,
          "id": "spectator-13",
          "isGuest": true,
          "name": "Orion_Ferry89",
        },
        {
          "canEdit": false,
          "id": "spectator-14",
          "isGuest": true,
          "name": "Jordan_Rutherford",
        },
        {
          "canEdit": false,
          "id": "spectator-15",
          "isGuest": true,
          "name": "Rosie.Daugherty",
        },
        {
          "canEdit": false,
          "id": "spectator-16",
          "isGuest": true,
          "name": "Damien55",
        },
        {
          "canEdit": false,
          "id": "spectator-17",
          "isGuest": true,
          "name": "Eleanora.Lehner88",
        },
        {
          "canEdit": false,
          "id": "spectator-18",
          "isGuest": true,
          "name": "Mable_Lakin",
        },
        {
          "canEdit": false,
          "id": "spectator-19",
          "isGuest": true,
          "name": "Axel.Ruecker",
        },
      ],
    },
    {
      "description": null,
      "document": {
        "content": "",
        "id": "document-1",
        "language": "TYPESCRIPT",
      },
      "id": "room-1",
      "name": "Inverse zero tolerance paradigm",
      "owner": {
        "displayName": "Hayley",
        "id": "user-1",
        "profilePicture": "https://via.placeholder.com/150x150/f35284/87e8eb.webp?text=Hayley",
      },
      "spectators": [],
    },
    {
      "description": null,
      "document": {
        "content": "",
        "id": "document-2",
        "language": "TYPESCRIPT",
      },
      "id": "room-2",
      "name": "Mandatory client-server adapter",
      "owner": {
        "displayName": "Hudson",
        "id": "user-2",
        "profilePicture": "https://via.placeholder.com/150x150/9a9773/4fd26a.jpeg?text=Hudson",
      },
      "spectators": [],
    },
    {
      "description": null,
      "document": {
        "content": "",
        "id": "document-3",
        "language": "TYPESCRIPT",
      },
      "id": "room-3",
      "name": "Optional analyzing orchestration",
      "owner": {
        "displayName": "Kathryne",
        "id": "user-3",
        "profilePicture": "https://via.placeholder.com/150x150/cd9d02/045129.png?text=Kathryne",
      },
      "spectators": [],
    },
    {
      "description": null,
      "document": {
        "content": "",
        "id": "document-4",
        "language": "TYPESCRIPT",
      },
      "id": "room-4",
      "name": "Versatile tertiary intranet",
      "owner": {
        "displayName": "Vito",
        "id": "user-4",
        "profilePicture": "https://via.placeholder.com/150x150/dfd43f/40f5d9.png?text=Vito",
      },
      "spectators": [],
    },
    {
      "description": null,
      "document": {
        "content": "",
        "id": "document-5",
        "language": "TYPESCRIPT",
      },
      "id": "room-5",
      "name": "Enhanced 6th generation strategy",
      "owner": {
        "displayName": "Damaris",
        "id": "user-5",
        "profilePicture": "https://via.placeholder.com/150x150/fa1d8a/7ba6f4.webp?text=Damaris",
      },
      "spectators": [],
    },
    {
      "description": null,
      "document": {
        "content": "",
        "id": "document-6",
        "language": "TYPESCRIPT",
      },
      "id": "room-6",
      "name": "Total value-added flexibility",
      "owner": {
        "displayName": "Sister",
        "id": "user-6",
        "profilePicture": "https://via.placeholder.com/150x150/82fd2b/d9b7ad.webp?text=Sister",
      },
      "spectators": [],
    },
    {
      "description": null,
      "document": {
        "content": "",
        "id": "document-7",
        "language": "TYPESCRIPT",
      },
      "id": "room-7",
      "name": "Centralized context-sensitive productivity",
      "owner": {
        "displayName": "Arden",
        "id": "user-7",
        "profilePicture": "https://via.placeholder.com/150x150/faf6b5/b11950.jpg?text=Arden",
      },
      "spectators": [],
    },
    {
      "description": null,
      "document": {
        "content": "",
        "id": "document-8",
        "language": "TYPESCRIPT",
      },
      "id": "room-8",
      "name": "Synergized 24/7 toolset",
      "owner": {
        "displayName": "Schuyler",
        "id": "user-8",
        "profilePicture": "https://via.placeholder.com/150x150/b590d9/a88bda.webp?text=Schuyler",
      },
      "spectators": [],
    },
    {
      "description": null,
      "document": {
        "content": "",
        "id": "document-9",
        "language": "TYPESCRIPT",
      },
      "id": "room-9",
      "name": "Advanced executive installation",
      "owner": {
        "displayName": "Adell",
        "id": "user-9",
        "profilePicture": "https://via.placeholder.com/150x150/26304d/935318.jpeg?text=Adell",
      },
      "spectators": [],
    },
    {
      "description": null,
      "document": {
        "content": "",
        "id": "document-10",
        "language": "TYPESCRIPT",
      },
      "id": "room-10",
      "name": "Robust disintermediate adapter",
      "owner": {
        "displayName": "Olen",
        "id": "user-10",
        "profilePicture": "https://via.placeholder.com/150x150/acac40/4fa535.gif?text=Olen",
      },
      "spectators": [],
    },
    {
      "description": null,
      "document": {
        "content": "",
        "id": "document-11",
        "language": "TYPESCRIPT",
      },
      "id": "room-11",
      "name": "Digitized actuating adapter",
      "owner": {
        "displayName": "Chelsie",
        "id": "user-11",
        "profilePicture": "https://via.placeholder.com/150x150/fe92a4/0dace2.webp?text=Chelsie",
      },
      "spectators": [],
    },
    {
      "description": null,
      "document": {
        "content": "",
        "id": "document-12",
        "language": "TYPESCRIPT",
      },
      "id": "room-12",
      "name": "Decentralized heuristic policy",
      "owner": {
        "displayName": "Cara",
        "id": "user-12",
        "profilePicture": "https://via.placeholder.com/150x150/f07e0d/601ff3.jpeg?text=Cara",
      },
      "spectators": [],
    },
    {
      "description": null,
      "document": {
        "content": "",
        "id": "document-13",
        "language": "TYPESCRIPT",
      },
      "id": "room-13",
      "name": "Robust mission-critical core",
      "owner": {
        "displayName": "Orion",
        "id": "user-13",
        "profilePicture": "https://via.placeholder.com/150x150/3fa9ad/dbbe8f.jpeg?text=Orion",
      },
      "spectators": [],
    },
    {
      "description": null,
      "document": {
        "content": "",
        "id": "document-14",
        "language": "TYPESCRIPT",
      },
      "id": "room-14",
      "name": "Open-architected upward-trending product",
      "owner": {
        "displayName": "Jordan",
        "id": "user-14",
        "profilePicture": "https://via.placeholder.com/150x150/364f4b/778bb4.png?text=Jordan",
      },
      "spectators": [],
    },
    {
      "description": null,
      "document": {
        "content": "",
        "id": "document-15",
        "language": "TYPESCRIPT",
      },
      "id": "room-15",
      "name": "Switchable secondary circuit",
      "owner": {
        "displayName": "Rosie",
        "id": "user-15",
        "profilePicture": "https://via.placeholder.com/150x150/c6eb06/c6f245.webp?text=Rosie",
      },
      "spectators": [],
    },
    {
      "description": null,
      "document": {
        "content": "",
        "id": "document-16",
        "language": "TYPESCRIPT",
      },
      "id": "room-16",
      "name": "Enhanced eco-centric interface",
      "owner": {
        "displayName": "Damien",
        "id": "user-16",
        "profilePicture": "https://via.placeholder.com/150x150/137f43/fe3d1c.webp?text=Damien",
      },
      "spectators": [],
    },
    {
      "description": null,
      "document": {
        "content": "",
        "id": "document-17",
        "language": "TYPESCRIPT",
      },
      "id": "room-17",
      "name": "Focused client-server internet solution",
      "owner": {
        "displayName": "Eleanora",
        "id": "user-17",
        "profilePicture": "https://via.placeholder.com/150x150/bb9e2e/4cd0c7.gif?text=Eleanora",
      },
      "spectators": [],
    },
    {
      "description": null,
      "document": {
        "content": "",
        "id": "document-18",
        "language": "TYPESCRIPT",
      },
      "id": "room-18",
      "name": "Profound analyzing installation",
      "owner": {
        "displayName": "Mable",
        "id": "user-18",
        "profilePicture": "https://via.placeholder.com/150x150/dc4a0e/9fc5c0.webp?text=Mable",
      },
      "spectators": [],
    },
    {
      "description": null,
      "document": {
        "content": "",
        "id": "document-19",
        "language": "TYPESCRIPT",
      },
      "id": "room-19",
      "name": "Compatible global process improvement",
      "owner": {
        "displayName": "Axel",
        "id": "user-19",
        "profilePicture": "https://via.placeholder.com/150x150/57e10e/2b0ffd.png?text=Axel",
      },
      "spectators": [],
    },
  ],
}
`;
